vulnerabilities/importers/github.py

```diff
@@ -0,0 +1,342 @@
+                        )
+        return adv_list
+
+    @staticmethod
+    def categorize_versions(
+        version_range: str, all_versions: Set[str]
+    ) -> Tuple[Set[str], Set[str]]:
+        version_range = RangeSpecifier(version_range)
+        affected_versions = {
+            version for version in all_versions if version in version_range
+        }
+        return (affected_versions, all_versions - affected_versions)
+
+
+class MavenVersionAPI:
+    def __init__(self):
+        self.cache = {}
+
+    def get(self, pkg_name: str) -> Set[str]:
+        return self.cache.get(pkg_name, set())
+
+    def load_to_api(self, pkg_name: str) -> None:
+
+        if pkg_name in self.cache:
+            return
+
+        artifact_comps = pkg_name.split(':')
+        endpoint = self.artifact_url(artifact_comps)
+        resp = requests.get(endpoint).content
+
+        try:
+
+            xml_resp = ET.ElementTree(ET.fromstring(resp.decode('utf-8')))
+            self.cache[pkg_name] = self.extract_versions(xml_resp)
+
+        except ET.ParseError:
+            self.cache[pkg_name] = set()
+
+    @staticmethod
+    def artifact_url(artifact_comps: List[str]) -> str:
+
+        base_url = 'https://repo.maven.apache.org/maven2/{}'
+        group_id, artifact_id = artifact_comps
+        group_url = group_id.replace('.', '/')
+        suffix = group_url + '/' + artifact_id + '/' + 'maven-metadata.xml'
+        endpoint = base_url.format(suffix)
+
+        return endpoint
+
+    @staticmethod
+    def extract_versions(xml_response: ET.ElementTree) -> Set[str]:
+
+        all_versions = set()
+        for child in xml_response.getroot().iter():
+            if child.tag == 'version':
+                all_versions.add(child.text)
+
+        return all_versions
+
+
+class NugetVersionAPI:
+    def __init__(self):
+        self.cache = {}
+
+    def get(self, pkg_name: str) -> Set[str]:
+        return self.cache.get(pkg_name.lower(), set())
+
+    def load_to_api(self, pkg_name: str) -> None:
+        if pkg_name in self.cache:
+            return
+        endpoint = self.nuget_url(pkg_name)
+        try:
+            resp = requests.get(endpoint).json()
+        # pkg_name=Microsoft.NETCore.UniversalWindowsPlatform triggers
+        # JSONDecodeError.
+        except json.decoder.JSONDecodeError:
+            self.cache[pkg_name.lower()] = set()
+            return
+
+        self.cache[pkg_name.lower()] = self.extract_versions(resp)
+
+    @staticmethod
+    def nuget_url(pkg_name: str) -> str:
+        base_url = 'https://api.nuget.org/v3/registration5-semver1/{}/index.json'
+        return base_url.format(pkg_name.lower())
+
+    @staticmethod
+    def extract_versions(json_resp: dict) -> Set[str]:
+        all_versions = set()
+        try:
+            for entry in json_resp['items'][0]['items']:
+                all_versions.add(entry['catalogEntry']['version'])
+        # json response for YamlDotNet.Signed triggers this exception
```

Comment:
What does this means exactly?What about using a cascade of get() instead to understand exactly why there are these key errors? Can you show me the problematic JSON?

Reply:
>Can you show me the problematic JSON?

This is raising that exception https://api.nuget.org/v3/registration5-semver1/yamldotnet.signed/index.json 
All others follow this schema instead https://api.nuget.org/v3/registration5-semver1/sustainsys.saml2/index.json 

>What about using a cascade of get() instead to understand exactly why there are these key errors? 

That can be done, but I thought catching an exception would be more 'pythonic' (EAFP).

Reply:
> That can be done, but I thought catching an exception would be more 'pythonic' (EAFP).

Well, the key is readability and clarity first: here when I read the code in `json_resp['items'][0]['items']` I wonder immediately. What if there is no items or this is not a mapping? What is [0]? Is this a list or a mapping with a key `0`? if this is a list, what about the other items in that list? and what is this second `items` about?

Here we could have something instead like that:
```
for item in json_resp.get('items', []):
    for entry in item.get('items', []):
        ...
```
When I read it I get these are mapping that return lists and I have no problem to understand the underlying data structure that I do not know about. Otherwise, I feel I need to do quite a bit of effort to read the code and it assumes that I know about the upstream JSON data structure before hand. So EAFP is fine, but it matters more to think about the readability of the code IMHO and how it will resist to future changes in the API and how that will make it easy to debug or not.



> This is raising that exception https://api.nuget.org/v3/registration5-semver1/yamldotnet.signed/index.json
> All others follow this schema instead https://api.nuget.org/v3/registration5-semver1/sustainsys.saml2/index.json

This is the kind of comment that is needed in the code :)

Reply:
I agree with everything you said. 

I am wondering what's the best answer in this context to the question you asked.

>how it will resist to future changes in the API and how that will make it easy to debug or not.

Because in this case, the end result in such an event, the outcome of both approaches  EAFP(try-except) and LBYL(get cascade)  is same.

Reply:
1. you can still keep inside your direct items access as `all_versions.add(entry['catalogEntry']['version'])` so if that part changes if will fail.
2. we should have tests too
3. we could think about a very limited subset of "canary"-like tests that are doing live queries on the web and fail if the API changes by returning different data.